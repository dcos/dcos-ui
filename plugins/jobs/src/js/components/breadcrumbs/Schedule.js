import React from "react";
import prettycron from "prettycron";
import PropTypes from "prop-types";
import { Tooltip } from "reactjs-components";
import { Icon } from "@dcos/ui-kit";
import { SystemIcons } from "@dcos/ui-kit/dist/packages/icons/dist/system-icons-enum";
import {
  greyDark,
  iconSizeXs
} from "@dcos/ui-kit/dist/packages/design-tokens/build/js/designTokens";

import BreadcrumbSupplementalContent from "#SRC/js/components/BreadcrumbSupplementalContent";

export default function ItemSchedule({ schedule }) {
  const { cron } = schedule;

  return (
    <BreadcrumbSupplementalContent>
      <Tooltip
        content={prettycron.toString(cron)}
        maxWidth={250}
        wrapText={true}
      >
        <Icon color={greyDark} shape={SystemIcons.Repeat} size={iconSizeXs} />
      </Tooltip>
    </BreadcrumbSupplementalContent>
  );
}

ItemSchedule.propTypes = {
  schedule: PropTypes.shape({
    enabled: PropTypes.bool.isRequired,
    cron: PropTypes.string.isRequired
  })
};
