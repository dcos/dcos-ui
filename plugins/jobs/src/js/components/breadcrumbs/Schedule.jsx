import React, { lazy, Suspense } from "react";
import PropTypes from "prop-types";
import { Icon } from "@dcos/ui-kit";
import { SystemIcons } from "@dcos/ui-kit/dist/packages/icons/dist/system-icons-enum";
import {
  textColorSecondary,
  iconSizeXs
} from "@dcos/ui-kit/dist/packages/design-tokens/build/js/designTokens";

import BreadcrumbSupplementalContent from "#SRC/js/components/BreadcrumbSupplementalContent";

const JobsCronTooltip = lazy(() =>
  import(
    /* webpackChunkName: "JobsCronTooltip" */ "#SRC/js/components/JobsCronTooltip"
  )
);

export default function ItemSchedule({ schedule }) {
  const { cron } = schedule;

  return (
    <BreadcrumbSupplementalContent>
      <Suspense
        fallback={
          <Icon
            shape={SystemIcons.Repeat}
            color={textColorSecondary}
            size={iconSizeXs}
          />
        }
      >
        <JobsCronTooltip content={cron} />
      </Suspense>
    </BreadcrumbSupplementalContent>
  );
}

ItemSchedule.propTypes = {
  schedule: PropTypes.shape({
    enabled: PropTypes.bool.isRequired,
    cron: PropTypes.string.isRequired
  })
};
