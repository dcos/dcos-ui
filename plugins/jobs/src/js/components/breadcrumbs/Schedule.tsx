import * as React from "react";
import PropTypes from "prop-types";
import { Icon } from "@dcos/ui-kit";
import { SystemIcons } from "@dcos/ui-kit/dist/packages/icons/dist/system-icons-enum";
import {
  textColorSecondary,
  iconSizeXs
} from "@dcos/ui-kit/dist/packages/design-tokens/build/js/designTokens";

import BreadcrumbSupplementalContent from "#SRC/js/components/BreadcrumbSupplementalContent";

const JobsCronTooltip = React.lazy(() =>
  import(
    /* webpackChunkName: "JobsCronTooltip" */ "#SRC/js/components/JobsCronTooltip"
  )
);

export default function ItemSchedule({ schedule }) {
  const { cron } = schedule;

  return (
    <BreadcrumbSupplementalContent>
      <React.Suspense
        fallback={
          <Icon
            shape={SystemIcons.Repeat}
            color={textColorSecondary}
            size={iconSizeXs}
          />
        }
      >
        <JobsCronTooltip content={cron} />
      </React.Suspense>
    </BreadcrumbSupplementalContent>
  );
}

ItemSchedule.propTypes = {
  schedule: PropTypes.shape({
    enabled: PropTypes.bool.isRequired,
    cron: PropTypes.string.isRequired
  })
};
