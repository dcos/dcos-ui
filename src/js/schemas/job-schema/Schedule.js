/* eslint-disable no-unused-vars */
import React from "react";
/* eslint-enable no-unused-vars */

import JobValidatorUtil from "../../utils/JobValidatorUtil";
import MetadataStore from "../../stores/MetadataStore";
import ValidatorUtil from "../../utils/ValidatorUtil";

const Schedule = {
  title: "Schedule",
  description: "Set time and date for the job to run",
  type: "object",
  properties: {
    runOnSchedule: {
      label: "Run on a schedule",
      showLabel: true,
      title: "Run on a schedule",
      type: "boolean",
      getter(job) {
        const [schedule] = job.getSchedules();

        return schedule != null;
      }
    },
    cron: {
      title: "CRON Schedule",
      helpBlock: (
        <span>
          Use cron format to set your schedule, e.g. <i>0 0 20 * *</i>
          {". "}
          <a
            href={MetadataStore.buildDocsURI("/deploying-jobs/")}
            target="_blank"
          >
            View documentation
          </a>.
        </span>
      ),
      type: "string",
      getter(job) {
        const [schedule = {}] = job.getSchedules();

        return schedule.cron;
      },
      externalValidator({ schedule }, definition) {
        if (!schedule.runOnSchedule) {
          return true;
        }

        if (!JobValidatorUtil.isValidCronSchedule(schedule.cron)) {
          definition.showError =
            "CRON Schedule must not be empty and it must " +
            "follow the correct CRON format specifications";

          return false;
        }

        return true;
      }
    },
    timezone: {
      title: "Time Zone",
      description: (
        <span>
          {"Enter time zone in "}
          <a
            href="http://www.timezoneconverter.com/cgi-bin/zonehelp"
            target="_blank"
          >
            TZ format
          </a>, e.g. America/New_York.
        </span>
      ),
      type: "string",
      getter(job) {
        const [schedule = {}] = job.getSchedules();

        return schedule.timezone;
      }
    },
    startingDeadlineSeconds: {
      title: "Starting Deadline",
      description:
        "Time in seconds to start the job if it misses " +
        "scheduled time for any reason. Missed jobs executions will be " +
        "counted as failed ones.",
      type: "number",
      getter(job) {
        const [schedule = {}] = job.getSchedules();

        return schedule.startingDeadlineSeconds;
      },
      externalValidator({ schedule }, definition) {
        if (!schedule.runOnSchedule) {
          return true;
        }

        if (!ValidatorUtil.isDefined(schedule.startingDeadlineSeconds)) {
          return true;
        }

        if (!ValidatorUtil.isNumberInRange(schedule.startingDeadlineSeconds)) {
          definition.showError = "Expecting a positive number here";

          return true;
        }

        return true;
      }
    },
    enabled: {
      label: "Enabled",
      showLabel: true,
      title: "Enabled",
      type: "boolean",
      getter(job) {
        const [schedule = {}] = job.getSchedules();

        return schedule.enabled !== undefined ? schedule.enabled : false;
      }
    }
  },
  required: ["cron"]
};

module.exports = Schedule;
