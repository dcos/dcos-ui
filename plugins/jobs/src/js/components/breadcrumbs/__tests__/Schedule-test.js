import React from "react";
import renderer from "react-test-renderer";

import Schedule from "../Schedule";

describe("Schedule", () => {
  it("renders the schedule", () => {
    expect(
      renderer
        .create(<Schedule schedule={{ cron: "0 0 20 * *", enabled: true }} />)
        .toJSON()
    ).toMatchSnapshot();
  });
});
