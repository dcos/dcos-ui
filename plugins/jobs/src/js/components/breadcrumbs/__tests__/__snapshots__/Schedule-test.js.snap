// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule renders the schedule 1`] = `
<div
  className="breadcrumb__content breadcrumb__content--supplemental"
>
  <div
    className="tooltip-wrapper text-align-center"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <svg
      aria-label="system-repeat icon"
      className="css-1vunjz0"
      height={16}
      preserveAspectRatio="xMinYMin meet"
      role="img"
      viewBox="0 0 16 16"
      width={16}
    >
      <use
        xlinkHref="#system-repeat"
      />
    </svg>
  </div>
</div>
`;
